adhocteam/Head-Start-TTADP
# CD: Add env variables to application [Part 2]

**Description of change**
In https://github.com/adhocteam/Head-Start-TTADP/pull/29 I added the env variables to the application.  However, the values were interpreted as string literals (not as the variable values); the deployed application had values like `AUTH_CLIENT_ID: ${SANDBOX_AUTH_CLIENT_ID}`.  

~Because circle ci doesn't read/interpret the deployment config files and transform the circleci project env variables, it's necessary to write the true values in "manually." (I used `sed`.)  I'm not exactly thrilled with this approach- as it leaves plain text secrets in the deployment env. The files/secrets ARE NOT copied onto the application instance though, because the deployment directory is included in .cfignore.~

~I tried two other strategies that proved unsuccessful.~
~1. Using `cf set-env` to pass env variables: This doesn't work well for us because it can only be used on a running application, then the application must be restaged for variables to take hold.  Given that nearly every sandbox deploy is not to a running application, prod has never been deployed, and I seem to be killed dev often, this doesn't seem like a good strategy for us.  I could always push, set-env, then restage but that isn't very efficient.~
~2. Using `cf push` flag `--var`: After a few hours of guess-and-check, I could not get the syntax of this to work. I can't find ANY examples on the internet of people using this to set environment variables (things like `instance` count and app `name` yes, but not true env variables.)~

@rahearn clued me in on how to correctly user the `--var` flag on `cf push` (strategy 2 in the above comments).  Variable substitution now works as follows. 
1. A secret is set in the circle ci environment variable list. `SANDBOX_AUTH_CLIENT_ID: secret`
2. The deployment pipeline invokes the `cf_depoy` command with the parameter `auth_client_id: SANDBOX_AUTH_CLIENT_ID`. 
3. Within the `cf deploy` command, `cf push` is called with the vars-file where `AUTH_CLIENT_ID: ((AUTH_CLIENT_ID))`.  On the same line, `--var` sets `((AUTH_CLIENT_ID))` equal to the `SANDBOX_AUTH_CLIENT_ID` circle ci environment variable.  The application manifest interprets all this and sets `AUTH_CLIENT_ID` equal to the `SANDBOX_AUTH_CLIENT_ID`.

This is better because secrets aren't being written to deployment config files. Before, .cfignore was the only thing keeping those secret laden deployment config files off of the application instance. This implementation keeps env variables where they should be, in the environment.

**How to test**
- check that most recent deployment on this branch passed https://app.circleci.com/pipelines/github/adhocteam/Head-Start-TTADP?branch=sj-fix-env-var
- check that variables are passed into sandbox application instance https://dashboard.fr.cloud.gov/applications/2oBn9LBurIXUNpfmtZCQTCHnxUM/150aceb1-bcd6-4a5d-961d-a8d8e1c1f1b5/variables
- check that sandbox application is running https://tta-smarthub-sandbox.app.cloud.gov/. After login, the redirect will fail with a 4xx (instead of it's previous 5xx error) until 

**Issue(s)**
* https://github.com/HHS/Head-Start-TTADP/issues/47

**Checklist**

- [n/a] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [x] Documentation updated - **_Added new readme to deployment vars dir, added section on secret management to main readme_**
    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified
